Fix part of #13687: Convert dicts to domain objects as a part of normalization through schema validation.

Nik-09 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #13687:
2. This PR does the following: Convert dicts to domain objects as a part of normalization through schema validation.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface in various display sizes (mainly phone, tablet, and desktop display size) to demonstrate that the changes made in this PR work correctly.
If the changes in your PRs are autogenerated via a script and you cannot provide proof for the changes then please leave a comment "No proof of changes needed because {{Reason}}".
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

Nik-09 committed:
trivial changes for creating pr.

Nik-09 commented:
After confirmation of these requirement I will update the **Overview** section of this PR.

This PR adds the following features.
1. Update dicts with domain objects as a part of normalization through schema validation.
2. Update the names of the arguments if we are transforming the dicts into domain_objects, since transformation is done in schema validation layer.
3. Transform domain objects validator file into payload validator.
4. Move the list of handlers class name constant into a seperate file (other than payload_validator) by creating a new one.

@rohitkatlaa @seanlip @vojtechjelinek please confirm these changes, so that I can start working on this PR.
Thank you.

Nik-09 commented:
This PR is also an extending part of these two PR's: 
1. https://github.com/oppia/oppia/pull/13456
2. https://github.com/oppia/oppia/pull/13422

seanlip commented:
Hi @Nik-09, thanks for creating this PR. Some notes:

- Please create an issue for this. That way, if this PR gets closed for some reason, the issue is still present. You can then change the title of this PR to "Fix #XXX".
- Could you please explain how you plan to update the names of the arguments? What mechanism will you use to do this? It might be worth outlining the possible options and stating which you think is best, perhaps in a doc.
- I don't really understand what points 3 and 4 are. Are these based on comments from other reviewers?

Thanks!

seanlip commented:
(Also @Nik-09 feel free to assign folks if you need a review or any action from them. The Assignee list refers to the "person/people who is supposed to take the next action in general", not just the next reviewer. Thanks!)

Nik-09 commented:
> Please create an issue for this. That way, if this PR gets closed for some reason, the issue is still present. You can then change the title of this PR to "Fix #XXX".

Created -- https://github.com/oppia/oppia/issues/13687

> Could you please explain how you plan to update the names of the arguments? What mechanism will you use to do this?

Started working on a doc.

> I don't really understand what points 3 and 4 are. Are these based on comments from other reviewers?

True, these changes are requested by @vojtechjelinek.

> (Also @Nik-09 feel free to assign folks if you need a review or any action from them. The Assignee list refers to the "person/people who is supposed to take the next action in general", not just the next reviewer. Thanks!)

Thank you @seanlip for your guidance.

Nik-09 committed:
fixes merge conflict

Nik-09 committed:
fixes merge conflict

seanlip commented:
Hi @Nik-09 -- just to check, what is the plan for this PR? (Just thinking that, if we want to set a convention we should probably do it soonish before people start following the incorrect thing.)

Thanks!

Nik-09 commented:
> Hi @Nik-09 -- just to check, what is the plan for this PR? (Just thinking that, if we want to set a convention we should probably do it soonish before people start following the incorrect thing.)
> 
> Thanks!

@seanlip Yes I was just facing a problem with my machine, I am working on a doc, will make this PR ready for review on or before 5th Sept. Thank you

seanlip commented:
Ah ok sounds good -- thanks @Nik-09!

Nik-09 committed:
Merge branch 'update_dict_to_domain_objects_as_normalization' of https://github.com/Nik-09/oppia into nikhil

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into update_dict_to_domain_objects_as_normalization

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into update_dict_to_domain_objects_as_normalization

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into update_dict_to_domain_objects_as_normalization

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into update_dict_to_domain_objects_as_normalization

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into update_dict_to_domain_objects_as_normalization

Nik-09 committed:
Removing changes for clarity.

Nik-09 committed:
Adds comment & changes method name

Nik-09 committed:
fixes backend tests

Nik-09 committed:
Fixes backend tests for schema utils.

Nik-09 commented on core/controllers/payload_validator_test.py:
```diff
@@ -66,11 +66,12 @@ def test_invalid_args_raises_exceptions(self) -> None:
-                allowed_extra_args=False,
-                allow_string_to_bool_conversion=False
+            normalized_value, errors = (
+                payload_validator.validate_arguments_against_schema(
```
Old method name was very generic so introduced a new name for the method which is more specific for Schema validation.

vojtechjelinek replied:
Yeah, good choice.

Nik-09 replied:
Thank you.

Nik-09 committed:
Updates docstring for domain objects validator file

Nik-09 committed:
Fixes lint issues

Nik-09 committed:
fixes lint issue

Nik-09 committed:
Introduced a feature for key modification by Schema validation.

Nik-09 committed:
updates docstring

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into update_dict_to_domain_objects_as_normalization

Nik-09 committed:
Conversion of dicts to domain object in controllers layer.

Nik-09 committed:
remove unused import

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into update_dict_to_domain_objects_as_normalization

Nik-09 committed:
Fixes backend tests.

Nik-09 commented:
@vojtechjelinek @DubeySandeep @nithusha21 @aks681 @vinitamurthi @BenHenning @seanlip 
Please take a look.
Thank you.

vojtechjelinek commented on core/controllers/domain_objects_validator.py:
```diff
@@ -38,19 +38,31 @@ def validate_exploration_change(obj):
     # No explicit call to validate_dict method is necessary, because
     # ExplorationChange calls validate method while initialization.
-    exp_domain.ExplorationChange(obj) # type: ignore[no-untyped-call]
+
```
Empty line not needed.

Nik-09 replied:
Removed

vojtechjelinek commented on core/controllers/domain_objects_validator.py:
```diff
@@ -38,19 +38,31 @@ def validate_exploration_change(obj):
+    # and in the handler these dict values are used to set config properties
+    # individually. Hence Conversion of dicts to domain objects is not required
+    # for new_config_properties.
+
```
Empty line not needed.

Nik-09 replied:
Removed

vojtechjelinek commented on core/controllers/domain_objects_validator.py:
```diff
@@ -60,14 +72,23 @@ def validate_new_config_property_values(obj):
+    # to update blog posts in the domain layer. This dict does not correspond
+    # to any domain class so we are validating the fields of change_dict
+    # as a part of schema validation.
+
```
Remove this empty line.

Nik-09 replied:
Removed

vojtechjelinek commented on core/controllers/domain_objects_validator.py:
```diff
@@ -84,27 +105,42 @@ def validate_change_dict_for_blog_post(change_dict):
-    validation_class = state_domain.State.from_dict(state_dict) # type: ignore[no-untyped-call]
-    validation_class.validate(None, True)
+    state_object = state_domain.State.from_dict(state_dict) # type: ignore[no-untyped-call]
+    state_object.validate(None, True)
```
Can you please use keywords for the `validate` function?

Nik-09 replied:
Added

vojtechjelinek commented on core/controllers/domain_objects_validator.py:
```diff
@@ -114,7 +150,15 @@ def validate_email_dashboard_data(
+    # query params. This dict represents the UserQueryParams class, which is a
+    # namedtuple. Hence the fields of the dict are being validated as a part of
+    # schema validation before saving new user queries in the handler.
+
```
Remove this empty line.

Nik-09 replied:
Removed

vojtechjelinek commented on core/controllers/domain_objects_validator.py:
```diff
@@ -124,14 +168,24 @@ def validate_email_dashboard_data(
+    # task_entry_dict. Thus, it is not possible to create the full
+    # domain object at the payload validation stage. Hence, the key-value pairs
+    # of task_entry_dict are being validated as a part of schema validation.
+
```
Remove this empty line.

Nik-09 replied:
Removed

vojtechjelinek commented on core/controllers/domain_objects_validator.py:
```diff
@@ -124,14 +168,24 @@ def validate_email_dashboard_data(
+    Returns:
+        dict. Returns the task entries dict after validation.
     """
+    # For creating the TaskEntry domain object, we have to include the
+    # exploration_id and the user_id which are not included in the
+    # task_entry_dict. Thus, it is not possible to create the full
+    # domain object at the payload validation stage. Hence, the key-value pairs
+    # of task_entry_dict are being validated as a part of schema validation.
```
This seems to be used a lot in this file. Did you verify that in any of these places we can do something to actually return the domain object.

Nik-09 replied:
Note: This validator is not in use as because all the backend calls to Improvements tab are temporarily removed.
As per your comment do I make a domain object here(param-validation layer) and return it, and in the handler layer update other fields so after it will form a complete object.
Please correct me if I am wrong at any moment.

vojtechjelinek replied:
Yes, I think your comment is correct.

Nik-09 replied:
Thank you.

vojtechjelinek commented on core/controllers/editor.py:
```diff
@@ -673,12 +667,12 @@ class StateYamlHandler(EditorHandler):
         self.render_json({
             'yaml': state_domain.State.convert_state_dict_to_yaml(
-                state_dict, width),
+                state_object, width),
```
Can we really use the object when we used dict previously?

Nik-09 replied:
Thank you, this argument is not sent to domain layer.
Only used by controllers, thus after validation instead of returning an object I have returned a dict.
Hence removed all the changes from here.

vojtechjelinek commented on core/controllers/editor.py:
```diff
@@ -1072,13 +1066,9 @@ def put(self, exploration_id):
-
+        change_list = self.normalized_payload.get('change_list')
         version = self.normalized_payload.get('version')
+
```
Why is this added?

Nik-09 replied:
Removed

vojtechjelinek commented on core/controllers/editor.py:
```diff
@@ -910,7 +904,7 @@ def post(self, exp_id):
         issue_to_remove = None
         for issue in exp_issues.unresolved_issues:
-            if issue.to_dict() == exp_issue_dict:
+            if issue.to_dict() == exp_issue_object.to_dict():
```
Can we do 
```suggestion
            if issue == exp_issue_object:
```

vojtechjelinek replied:
You will probably need to introduce `__eq__` into the `ExplorationIssue` class.

Nik-09 replied:
Thank you, this was something new.
Updated.

vojtechjelinek commented on core/controllers/payload_validator.py:
```diff
@@ -27,15 +27,41 @@
 from typing import Any, Dict, List, Optional, Tuple, Union
 
 
+def get_schema_type(arg_schema: Dict[str, Any]) -> Any:
```
```suggestion
def get_schema_type(arg_schema: Dict[str, Any]) -> str:
```

Also please add a comment explaining why the `arg_schema` is `Dict[str, Any]`

Nik-09 replied:
Added.

vojtechjelinek commented on core/controllers/payload_validator.py:
```diff
@@ -27,15 +27,41 @@
 
+def get_schema_type(arg_schema: Dict[str, Any]) -> Any:
+    """This method returns the schema type for an argument by accepting the
+    schema for the corresponding argument.
```
```suggestion
    """Returns the schema type for an argument by accepting the
    schema for the corresponding argument.
```

Also, I don't understand this sentence.

Nik-09 replied:
Updated.

vojtechjelinek commented on core/controllers/payload_validator.py:
```diff
@@ -27,15 +27,41 @@
+    return arg_schema['schema']['type']
+
+
+def get_corresponding_key_for_object(arg_schema: Dict[str, Any]) -> Any:
```
```suggestion
def get_corresponding_key_for_object(arg_schema: Dict[str, Any]) -> str:
```

Again, please add an explanatory comment.

Nik-09 replied:
Updated

vojtechjelinek commented on core/controllers/payload_validator.py:
```diff
@@ -27,15 +27,41 @@
+
+
+def get_corresponding_key_for_object(arg_schema: Dict[str, Any]) -> Any:
+    """This method extracts the new key for an argument from its schema.
```
```suggestion
    """Returns the new key for an argument from its schema.
```

Nik-09 replied:
Updated

vojtechjelinek commented on core/controllers/payload_validator.py:
```diff
@@ -27,15 +27,41 @@
+        arg_schema: dict(str, *). Schema for an argument.
+
+    Returns:
+        str. Returns the new argument name as mentioned in the schema.
```
```suggestion
        str. The new argument name.
```

Nik-09 replied:
Updated

vojtechjelinek commented on core/controllers/payload_validator.py:
```diff
@@ -27,15 +27,41 @@
         allowed_extra_args: bool,
         allow_string_to_bool_conversion: bool = False
-) -> Tuple[Dict[str, Any], List[str]]:
+    ) -> Tuple[Dict[str, Any], List[str]]:
```
```suggestion
) -> Tuple[Dict[str, Any], List[str]]:
```

Nik-09 replied:
Updated.

vojtechjelinek commented on core/controllers/payload_validator.py:
```diff
@@ -27,15 +27,41 @@
         handler_args: Any,
         handler_args_schemas: Any,
         allowed_extra_args: bool,
         allow_string_to_bool_conversion: bool = False
```
You can shift these 4 spaces to left.

Nik-09 replied:
Updated

vojtechjelinek commented on core/controllers/payload_validator.py:
```diff
@@ -27,15 +27,41 @@
         allow_string_to_bool_conversion: bool = False
-) -> Tuple[Dict[str, Any], List[str]]:
+    ) -> Tuple[Dict[str, Any], List[str]]:
+
```
Why is this added?

Nik-09 replied:
Updated.

vojtechjelinek commented on core/controllers/payload_validator.py:
```diff
@@ -53,16 +79,22 @@ def validate(
     """
     # Collect all errors and present them at once.
     errors = []
+    # Dictionary to hold normalizad values of arguments after validation.
```
```suggestion
    # Dictionary to hold normalized values of arguments after validation.
```

Nik-09 replied:
Thank you, fixed

vojtechjelinek commented on core/controllers/payload_validator.py:
```diff
@@ -53,16 +79,22 @@ def validate(
+                    'default_value' in arg_schema and
+                    arg_schema['default_value'] is None
+            ):
+                # Skip validation because the argument is optional.
```
```suggestion
            if 'default_value' in arg_schema:
                if arg_schema['default_value'] is None: 
                    # Skip validation because the argument is optional.
```

Nik-09 replied:
Modified.

vojtechjelinek commented on core/controllers/payload_validator.py:
```diff
@@ -53,16 +79,22 @@ def validate(
+            elif (
+                    'default_value' in arg_schema and
+                    arg_schema['default_value'] is not None
+            ):
```
```suggestion
                elif arg_schema['default_value'] is not None:
```

Nik-09 replied:
Updated

vojtechjelinek commented on core/controllers/payload_validator.py:
```diff
@@ -53,16 +79,22 @@ def validate(
+                    arg_schema['default_value'] is not None
+            ):
                 handler_args[arg_key] = arg_schema['default_value']
             elif 'default_value' not in arg_schema:
```
```suggestion
            else:
```

Nik-09 replied:
Updated.

vojtechjelinek commented on core/controllers/payload_validator.py:
```diff
@@ -72,12 +104,17 @@ def validate(
+        # Modification of argument name if new_key_for_argument field is present
+        # in the schema.
+        if 'new_key_for_argument' in arg_schema['schema']:
+            arg_key = get_corresponding_key_for_object(arg_schema)
```
Is this used in any existing handler? If not can you introduce it somewhere?

Nik-09 replied:
Introduced in ```ResolveIssueHandler```

vojtechjelinek commented on core/controllers/payload_validator_test.py:
```diff
@@ -66,11 +66,12 @@ def test_invalid_args_raises_exceptions(self) -> None:
+                    handler_args,
+                    handler_args_schema,
+                    allowed_extra_args=False,
+                    allow_string_to_bool_conversion=False)
```
```suggestion
                    allow_string_to_bool_conversion=False
                )
```

Nik-09 replied:
Updated

vojtechjelinek commented on core/controllers/payload_validator_test.py:
```diff
@@ -124,11 +125,12 @@ def test_valid_args_do_not_raises_exception(self) -> None:
+                    handler_args,
+                    handler_args_schema,
+                    allowed_extra_args=False,
+                    allow_string_to_bool_conversion=True)
```
```suggestion
                    allow_string_to_bool_conversion=True
                )
```

Nik-09 replied:
Updated

vojtechjelinek commented on core/controllers/payload_validator_test.py:
```diff
@@ -148,3 +150,35 @@ def test_convert_string_to_bool(self) -> None:
+class CheckGetSchemaTypeMethod(test_utils.GenericTestBase):
+    """Test class to check behaviour of get_schema_type method."""
+
+    def test_check_new_arg_key(self) -> None:
```
Wrong name.

Nik-09 replied:
Modified

vojtechjelinek commented on core/domain/state_domain.py:
```diff
@@ -2673,29 +2673,18 @@ def can_undergo_classification(self):
 
     @classmethod
-    def convert_state_dict_to_yaml(cls, state_dict, width):
+    def convert_state_dict_to_yaml(cls, state_object, width):
```
Change the name of the function.

Nik-09 replied:
Removed the changes, not required after modification.

vojtechjelinek commented on core/domain/state_domain.py:
```diff
@@ -2673,29 +2673,18 @@ def can_undergo_classification(self):
 
         Args:
-            state_dict: dict. A dict representing a state in an exploration.
+            state_object: State. State obje in an exploration.
```
```suggestion
            state_object: State. State in an exploration.
```

Nik-09 replied:
Removed, not required.

vojtechjelinek reviewed (changes requested):
Thanks! Left a few comments.

Nik-09 committed:
Removed extra spaces.

Nik-09 committed:
requested changes.

Nik-09 committed:
fixes merge conflict

Nik-09 committed:
fixing lint issue

Nik-09 committed:
Added comment.

Nik-09 committed:
Adds comment.

Nik-09 committed:
Fixes merge conflict

vojtechjelinek commented on core/controllers/domain_objects_validator.py:
```diff
@@ -37,19 +37,29 @@ def validate_exploration_change(obj):
-    for (name, value) in obj.items():
+    # The new_config_property values do not represent a domain class directly
+    # and in the handler these dict values are used to set config properties
+    # individually. Hence Conversion of dicts to domain objects is not required
```
```suggestion
    # individually. Hence conversion of dicts to domain objects is not required
```

vojtechjelinek replied:
Maybe move this above the `return`?

Nik-09 replied:
Updated.

vojtechjelinek commented on core/controllers/domain_objects_validator.py:
```diff
@@ -83,27 +101,46 @@ def validate_change_dict_for_blog_post(change_dict):
     # No explicit call to validate_dict method is necessary, because
     # CollectionChange calls validate method while initialization.
-    collection_domain.CollectionChange(obj) # type: ignore[no-untyped-call]
+
+    # Object should not be returned from here because it requires modification
+    # in many of the methods in the domain layer of the codebase and we are
+    # planning to remove collections from our codebase hence modification is
+    # not done here.
+
+    collection_domain.CollectionChange(collection_change_dict)
+    return collection_change_dict
```
```suggestion
    # Object should not be returned from here because it requires modification
    # in many of the methods in the domain layer of the codebase and we are
    # planning to remove collections from our codebase hence modification is
    # not done here.
    collection_domain.CollectionChange(collection_change_dict)
    return collection_change_dict
```

Nik-09 replied:
Updated.

vojtechjelinek commented on core/controllers/domain_objects_validator.py:
```diff
@@ -59,14 +69,22 @@ def validate_new_config_property_values(obj):
+    Returns:
+        dict. Returns the change_dict after validation.
     """
+    # The method returns a dict containing blog post properties, they are used
+    # to update blog posts in the domain layer. This dict does not correspond
+    # to any domain class so we are validating the fields of change_dict
+    # as a part of schema validation.
```
Maybe move this above the `return`?

Nik-09 replied:
Updated.

vojtechjelinek commented on core/controllers/domain_objects_validator.py:
```diff
@@ -152,9 +206,14 @@ def validate_aggregated_stats(aggregated_stats):
     Raises:
         InvalidInputException. Property not in aggregated stats dict.
     """
+    # The aggregated_stats parameter do not represents any domain class, hence
+    # dict form of the data is returned from here.
```
Move this below.

vojtechjelinek commented on core/controllers/payload_validator.py:
```diff
@@ -26,14 +26,42 @@
+
+
+# Here Dict[str, Any] is used for arg_schema because the value field of the
+# schema is itself a dict which furthur contains several nested dicts.
```
```suggestion
# schema is itself a dict that can further contain several nested dicts.
```

Nik-09 replied:
Updated.

vojtechjelinek commented on core/controllers/payload_validator.py:
```diff
@@ -26,14 +26,42 @@
 
 
+# Here Dict[str, Any] is used for arg_schema because the value field of the
+# schema is itself a dict which furthur contains several nested dicts.
```
```suggestion
# schema is itself a dict that can further contain several nested dicts.
```

Nik-09 replied:
Updated

vojtechjelinek commented on core/controllers/payload_validator.py:
```diff
@@ -52,31 +80,41 @@ def validate(
+        # Modification of argument name if new_key_for_argument field is present
+        # in the schema.
+        if 'new_key_for_argument' in arg_schema['schema']:
+            old_arg_key = arg_key
+            arg_key = get_corresponding_key_for_object(arg_schema)
+            handler_args[arg_key] = handler_args.pop(old_arg_key)
+            handler_args_schemas[arg_key] = (
+                handler_args_schemas.pop(old_arg_key))
```
```suggestion
            new_arg_key = get_corresponding_key_for_object(arg_schema)
            handler_args[new_arg_key] = handler_args.pop(arg_key)
            handler_args_schemas[new_arg_key] = (
                handler_args_schemas.pop(arg_key))
```

Nik-09 replied:
Updated

vojtechjelinek commented on core/domain/stats_domain.py:
```diff
@@ -876,6 +876,17 @@ def __init__(
         self.schema_version = schema_version
         self.is_valid = is_valid
 
+    def __eq__(self, other):
+        if isinstance(other, ExplorationIssue):
+            return (
+                self.issue_type == other.issue_type and
+                self.issue_customization_args == other.issue_customization_args
+                and self.playthrough_ids == other.playthrough_ids
+                and self.schema_version == other.schema_version
+                and self.is_valid == other.is_valid
+            )
+        return False
```
You might need to introduce `__hash__` too.

Nik-09 replied:
Implicit calling of ``` __hash__ ```while using ```object1 == object2``` is not done. (where object1 and object2 are the objects of same class) Hence in this class ``` __eq__()``` is only used.

vojtechjelinek reviewed (changes requested):
Thanks! Left a few comments.

BenHenning commented on core/domain/platform_feature_services.py:
```diff
@@ -114,15 +114,15 @@ def is_feature_enabled(feature_name):
         commit_message: str. The commit message.
-        new_rule_dicts: list(dict). A list of dict mappings of all fields
-            of PlatformParameterRule object.
+        new_rules: list(PlatformParameterRule). A list of dict of
```
This looks like this might need to be updated a bit more; this is no longer a list of dictionaries.

Perhaps instead: ``A list of PlatformParameterRule objects to update.`` might be clearer?

Nik-09 replied:
Updated.

BenHenning commented on core/domain/platform_parameter_registry.py:
```diff
@@ -151,31 +151,28 @@ def get_platform_parameter(cls, name):
-            new_rule_dicts: list(dist). A list of dict mappings of all fields
-                of PlatformParameterRule object.
+            new_rules: list(PlatformParameterRule). A list of
+                PlatformParameterRule object.
```
Suggest pluralizing to ``objects`` since the current grammar is a bit off.

Nik-09 replied:
Updated.

BenHenning reviewed (commented):
Thanks @Nik-09! PR LGTM for codeowners, but I had 2 docstring update comments. PTAL.

aks681 reviewed (approved):
Lgtm for 2 codeowner files

seanlip reviewed (approved):
Thanks @Nik-09, no concerns for my 2 codeowner files.

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into update_dict_to_domain_objects_as_normalization

DubeySandeep reviewed (approved):
LGTM for the changes in `core/schema_utils.py` and admin.py file

Nik-09 committed:
Updates comments and docstrings

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into update_dict_to_domain_objects_as_normalization

Nik-09 committed:
Fixes lint issues

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into update_dict_to_domain_objects_as_normalization

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into update_dict_to_domain_objects_as_normalization

Nik-09 committed:
fixes backend tests.

Nik-09 committed:
Fixes backend coverage issue.

Nik-09 committed:
Fixes merge conflict.

Nik-09 committed:
Fixes backend coverage.

Nik-09 committed:
Fixes merge conflicts.

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into update_dict_to_domain_objects_as_normalization

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into update_dict_to_domain_objects_as_normalization

Nik-09 committed:
Fixes backend coverage.

srijanreddy98 committed:
Fixes part of #9749: Migrate few instances of angular-html-bind (#14263)

* Revert "Revert "Fix part of #9749: Create a component that shows rte-output-display and use it instead of angular-html-bind. (#13229)" (#13361)"

This reverts commit 792de4596996527a696d0041e0eff2b8384eda30.

* Fix issues

* Fix spelling mistakes in comment

* Merge with upstream/develop

* Address review comments:
  1. Move oppia-rte-output.spec from the list of ignored files for innerHTML check in eslintrc and move to inline disables.
  2. Change the CODEOWNER listing from oppia-rte-parser.service.ts to oppia-rte-parser.service*.ts to include the spec file.
  3. Add refernce to the note at the top of file for all value.replace ops.
  4. Miscellaneous changes.

srijanreddy98 committed:
Fixes part of #9749: Migrate few instances of angular-html-bind (#14263)

* Revert "Revert "Fix part of #9749: Create a component that shows rte-output-display and use it instead of angular-html-bind. (#13229)" (#13361)"

This reverts commit 792de4596996527a696d0041e0eff2b8384eda30.

* Fix issues

* Fix spelling mistakes in comment

* Merge with upstream/develop

* Address review comments:
  1. Move oppia-rte-output.spec from the list of ignored files for innerHTML check in eslintrc and move to inline disables.
  2. Change the CODEOWNER listing from oppia-rte-parser.service.ts to oppia-rte-parser.service*.ts to include the spec file.
  3. Add refernce to the note at the top of file for all value.replace ops.
  4. Miscellaneous changes.

srijanreddy98 committed:
Fixes part of #9749: Migrate few instances of angular-html-bind (#14263)

* Revert "Revert "Fix part of #9749: Create a component that shows rte-output-display and use it instead of angular-html-bind. (#13229)" (#13361)"

This reverts commit 792de4596996527a696d0041e0eff2b8384eda30.

* Fix issues

* Fix spelling mistakes in comment

* Merge with upstream/develop

* Address review comments:
  1. Move oppia-rte-output.spec from the list of ignored files for innerHTML check in eslintrc and move to inline disables.
  2. Change the CODEOWNER listing from oppia-rte-parser.service.ts to oppia-rte-parser.service*.ts to include the spec file.
  3. Add refernce to the note at the top of file for all value.replace ops.
  4. Miscellaneous changes.

Co-authored-by: Vasa Srijan Reddy <redvasa@amazon.com>

seanlip committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

seanlip committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

seanlip committed:
Fixes part of #9749: Migrate few instances of angular-html-bind (#14263)

* Revert "Revert "Fix part of #9749: Create a component that shows rte-output-display and use it instead of angular-html-bind. (#13229)" (#13361)"

This reverts commit 792de45.

* Fix issues

* Fix spelling mistakes in comment

* Merge with upstream/develop

* Address review comments:
  1. Move oppia-rte-output.spec from the list of ignored files for innerHTML check in eslintrc and move to inline disables.
  2. Change the CODEOWNER listing from oppia-rte-parser.service.ts to oppia-rte-parser.service*.ts to include the spec file.
  3. Add refernce to the note at the top of file for all value.replace ops.
  4. Miscellaneous changes.

seanlip committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

seanlip committed:
Fixes part of #9749: Migrate few instances of angular-html-bind (#14263)
* Revert "Revert "Fix part of #9749: Create a component that shows rte-output-display and use it instead of angular-html-bind. (#13229)" (#13361)"

This reverts commit 792de45.

* Fix issues

* Fix spelling mistakes in comment

* Merge with upstream/develop

* Address review comments:
  1. Move oppia-rte-output.spec from the list of ignored files for innerHTML check in eslintrc and move to inline disables.
  2. Change the CODEOWNER listing from oppia-rte-parser.service.ts to oppia-rte-parser.service*.ts to include the spec file.
  3. Add refernce to the note at the top of file for all value.replace ops.
  4. Miscellaneous changes.

seanlip committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into update_dict_to_domain_objects_as_normalization

srijanreddy98 committed:
Merge branch 'develop' into update_dict_to_domain_objects_as_normalization

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into update_dict_to_domain_objects_as_normalization

Nik-09 committed:
Merge branch 'update_dict_to_domain_objects_as_normalization' of https://github.com/Nik-09/oppia into update_dict_to_domain_objects_as_normalization

Nik-09 committed:
Removes __hash__() method from ExplorationIssue class.

Nik-09 reviewed (commented):
@vojtechjelinek, @BenHenning I have updated the PR with suggested changes.
PTAL
Thank you.

vojtechjelinek commented on core/controllers/payload_validator_test.py:
```diff
@@ -114,20 +116,23 @@ def test_valid_args_do_not_raises_exception(self) -> None:
                     'schema': {
-                        'type': 'bool'
+                        'type': 'bool',
+                        'new_key_for_argument': 'new_kay_for_apply_draft'
```
```suggestion
                        'new_key_for_argument': 'new_key_for_apply_draft'
```

Nik-09 replied:
Updated.

vojtechjelinek commented on core/controllers/payload_validator_test.py:
```diff
@@ -114,20 +116,23 @@ def test_valid_args_do_not_raises_exception(self) -> None:
                 }
             }, {
-                'apply_draft': True
+                'new_kay_for_apply_draft': True
```
```suggestion
                'new_key_for_apply_draft': True
```

Nik-09 replied:
Updated

vojtechjelinek commented on core/domain/platform_feature_services.py:
```diff
@@ -113,15 +113,15 @@ def is_feature_enabled(feature_name):
 
 def update_feature_flag_rules(
-        feature_name, committer_id, commit_message, new_rule_dicts):
+        feature_name, committer_id, commit_message, new_rules):
```
```suggestion
    feature_name, committer_id, commit_message, new_rules
):
```

Nik-09 replied:
Updated.

vojtechjelinek commented on core/domain/platform_parameter_registry.py:
```diff
@@ -150,31 +150,28 @@ def get_platform_parameter(cls, name):
     @classmethod
     def update_platform_parameter(
-            cls, name, committer_id, commit_message, new_rule_dicts):
+            cls, name, committer_id, commit_message, new_rules):
```
```suggestion
        cls, name, committer_id, commit_message, new_rules
    ):
```

Nik-09 replied:
Updated.

vojtechjelinek commented on core/domain/stats_domain.py:
```diff
@@ -875,6 +875,15 @@ def __init__(
 
+    def __eq__(self, other):
+        return (
+            self.issue_type == other.issue_type and
+            self.issue_customization_args == other.issue_customization_args
+            and self.playthrough_ids == other.playthrough_ids
+            and self.schema_version == other.schema_version
+            and self.is_valid == other.is_valid
```
```suggestion
            self.issue_type == other.issue_type and
            self.issue_customization_args == other.issue_customization_args and
            self.playthrough_ids == other.playthrough_ids and
            self.schema_version == other.schema_version and
            self.is_valid == other.is_valid
```

Nik-09 replied:
Updated.

vojtechjelinek commented on core/domain/stats_domain_test.py:
```diff
@@ -1113,6 +1113,35 @@ def test_validate_playthrough_id_type(self):
+            2,
+            True
+        )
+        exp_issue3 = exp_issue1
```
This is a reference assignment, there is no chance that this object will be different. To properly check the method create a new object but with new values.

Nik-09 replied:
Updated.

vojtechjelinek reviewed (changes requested):
Thanks! Left a few last comments.

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into update_dict_to_domain_objects_as_normalization

Nik-09 committed:
Addresses suggested changes

Nik-09 committed:
Fixes merge conflicts from upstream/develop.

Nik-09 committed:
Fixes lint issues.

Nik-09 reviewed (commented):
@vojtechjelinek I have addressed all of your review comments and made modifications based on it. PTAL.

Thank you.

vojtechjelinek commented on core/domain/platform_feature_services.py:
```diff
@@ -113,15 +113,15 @@ def is_feature_enabled(feature_name):
 
 def update_feature_flag_rules(
-        feature_name, committer_id, commit_message, new_rule_dicts):
+    feature_name, committer_id, commit_message, new_rules):
```
```suggestion
    feature_name, committer_id, commit_message, new_rules
):
```

Nik-09 replied:
Done

vojtechjelinek commented on core/domain/platform_parameter_registry.py:
```diff
@@ -150,31 +150,28 @@ def get_platform_parameter(cls, name):
     @classmethod
     def update_platform_parameter(
-            cls, name, committer_id, commit_message, new_rule_dicts):
+        cls, name, committer_id, commit_message, new_rules):
```
```suggestion
        cls, name, committer_id, commit_message, new_rules
    ):
```

Nik-09 replied:
Done

vojtechjelinek reviewed (approved):
Thanks! Two small nits.

Nik-09 committed:
Addresses review comments
